open-craft/openedx-completion-aggregator
# Add indexes to StaleCompletion table

**Description:** This adds indexes to the stale completion table.  They cannot be added the simple way, because that locks the table for multiple minutes.  So instead, we perform the following steps:

1. Create a new table with the indexes.
2. Copy the data to the table in batches (up to a known maximum PK).
3. Rename the tables. 
4. Copy any data with a PK higher than the original maximum, again in batches, just to be safe.

Batch size is set to 1000 rows, which makes each batch take about half a second, which keeps lock contention low.

This has been tested on a data set of 1,000,000 rows, and took on the order of 8 minutes.  , so it will probably take around four hours in staging, but will not hold a lock on the tables during that time.

Note: The procedure for migrating the data without downtime was cribbed freely from the excellent Percona tools: https://www.percona.com/doc/percona-toolkit/2.2/pt-online-schema-change.html

**JIRA:** YONK-1175

**Dependencies:** 

**Merge deadline:** ASAP

**Installation instructions:** 

**Testing instructions:**

Verify that the index does its job.  querying the completion_aggregator_stalecompletion table WHERE username=... AND course_key=... AND resolved=False AND created < NOW() does not perform a full table scan. 

Verify that the migration runs, adds indexes to the table, and does not prevent data from being added to the table while it is running:

```sql
INSERT INTO completion_aggregator_stalecompletion 
    (created, modified, username, course_key, block_key, `force`, resolved)
    VALUES 
        (NOW(), NOW(), "username", "COURSE/COURSE/COURSE", "i4x:COURSE/COURSE/html/BLOCK", FALSE, FALSE)
```

should be sufficient to create a record while the migration is running.

**Reviewers:**

**Merge checklist:**

**Post merge:**
      finished.
